### PR TITLE
Ctpdev: RawToDigit

### PR DESCRIPTION
**Pull request by lietava**
Hi @shahor02 , I started to work on DigitToRaw. 
0I tried to follow the CPV example you suggested in previous PR. I added files in CTP workfloa and now trying to modify them. 
I have couple of questions:
1.) as we discussed CTP has no MC truth so I can ignore all related staff, right ?
2.)  main question:
void RawToDigitConverterSpec::run(framework::ProcessingContext& ctx) calls o2::xxx::RawReaderMemory rawreader(o2::framework::DataRefUtils::as<const char>(rawData));
which data like in DMA memory : I guess data created by o2::ctp::Digits2Raw are in that format ?
Should I create also RawReadrMemory like in CPV ?
Cheers, R.

### CODE REVIEW

**Comment by shahor02** [comment]
Hi @lietava 

> 1.) as we discussed CTP has no MC truth so I can ignore all related staff, right ?

Right, I don't see what can be the MC truth for the CTP. One can think of the presence of hadronic interaction (the alternative being generating trigger on QED or some late particle/afterpulse), but this can be checked directly with the info we have in the AOD.

> 2.) main question:
> void RawToDigitConverterSpec::run(framework::ProcessingContext& ctx) calls o2::xxx::RawReaderMemory rawreader(o2::framework::DataRefUtils::as(rawData));
> which data like in DMA memory : I guess data created by o2::ctp::Digits2Raw are in that format ?

Yes, o2-raw-file-reader injects into the the DPL the the raw data for each link: RDH followed by the payload you have provided when creating raw data.

> Should I create also RawReadrMemory like in CPV ?

There is no any convention on naming the routines, but in one way or another you should navigate over the inputs of `CTP/RAWDATA` and convert raw data to digit. You can also check how I do this and cache the data in ITS/TPC decoder:
https://github.com/AliceO2Group/AliceO2/blob/3ad766bacf1513a9cf9dea485a9c02b95d8621a8/Detectors/ITSMFT/common/workflow/src/STFDecoderSpec.cxx#L110
https://github.com/AliceO2Group/AliceO2/blob/3ad766bacf1513a9cf9dea485a9c02b95d8621a8/Detectors/ITSMFT/common/reconstruction/src/RawPixelDecoder.cxx#L125

Cheers,
 Ruben

**Comment by lietava** [comment]
> Hi @lietava
> 
> > 1.) as we discussed CTP has no MC truth so I can ignore all related staff, right ?
> 
> Right, I don't see what can be the MC truth for the CTP. One can think of the presence of hadronic interaction (the alternative being generating trigger on QED or some late particle/afterpulse), but this can be checked directly with the info we have in the AOD.
> 
> > 2.) main question:
> > void RawToDigitConverterSpec::run(framework::ProcessingContext& ctx) calls o2::xxx::RawReaderMemory rawreader(o2::framework::DataRefUtils::as(rawData));
> > which data like in DMA memory : I guess data created by o2::ctp::Digits2Raw are in that format ?
> 
> Yes, o2-raw-file-reader injects into the the DPL the the raw data for each link: RDH followed by the payload you have provided when creating raw data.
> 
> > Should I create also RawReadrMemory like in CPV ?
> 
> There is no any convention on naming the routines, but in one way or another you should navigate over the inputs of `CTP/RAWDATA` and convert raw data to digit. You can also check how I do this and cache the data in ITS/TPC decoder:
> https://github.com/AliceO2Group/AliceO2/blob/3ad766bacf1513a9cf9dea485a9c02b95d8621a8/Detectors/ITSMFT/common/workflow/src/STFDecoderSpec.cxx#L110
> 
> 
> https://github.com/AliceO2Group/AliceO2/blob/3ad766bacf1513a9cf9dea485a9c02b95d8621a8/Detectors/ITSMFT/common/reconstruction/src/RawPixelDecoder.cxx#L125
> 
> Cheers,
> Ruben

HI @shahor02 , sorry, I am still a bit lost - conceptually.
1.) input : I have a raw data file on disk : in your examples there is no file input. Do I get raw input  from workflow or I can do both  and workflow recognise it ?
2.) Decoding: Then when I get raw data : my case is probably simple. Is there a general method which gives me TF (RDH + Payload) ?
In that case everything is quite simple. Why I need to buffer anything ?
Cheers, Roman.

**Comment by shahor02** [comment]
Hi @lietava 

> 1.) input : I have a raw data file on disk : in your examples there is no file input. Do I get raw input from workflow or I can do both and workflow recognise it ?

The simulated raw data can be injected into the DPL workflow by the `o2-raw-file-reader`, see https://github.com/AliceO2Group/AliceO2/tree/dev/Detectors/Raw#raw-data-file-reader-workflow
For the example of the complex workflow starting with this reader see e.g. https://github.com/AliceO2Group/AliceO2/blob/dev/prodtests/full-system-test/dpl-workflow.sh, which is started from the full system test https://github.com/AliceO2Group/AliceO2/blob/dev/prodtests/full_system_test.sh

> 2.) Decoding: Then when I get raw data : my case is probably simple. Is there a general method which gives me TF (RDH + Payload) ?
> In that case everything is quite simple. Why I need to buffer anything ?

The examples I've referred to do exactly this: you extract from the input the messages containing your RDH + payload. E.g.
````
void CTPDecoder::run(ProcessingContext& pc) 
{
  std::vector<InputSpec> filter{InputSpec{"filter", ConcreteDataTypeMatcher{"CTP", "RAWDATA"}, Lifetime::Timeframe}};
  
  o2::framework::DPLRawParser parser(pc.inputs(), filter);
  for (auto it = parser.begin(); it != parser.end(); ++it) {
     auto rdh = it.get_if<o2::header::RAWDataHeader>();
     gsl::span<const uint8_t> payload(it.data(), it.size());
    // process RDH and payload
  }
````
allows you to iterate over all raw pages (a single link message may have multiple pages).
Alternatively, you can iterate over the input messages using
````
auto walker = o2::framework::InputRecordWalker(pc.inputs(), filter);
for (auto it = walker.begin(); it != walker.end(); ++it) {
  auto rawspan = o2::framework::DataRefUtils::as<const char>(it);
  /// here you get the gsl::span<const char> and should process as you would do with data in DMA,
  /// parsing the RDHs and navigating to next one yourself.
}
````

**Comment by lietava** [comment]
@shahor02 Thanks for example above. Please, check if all looks reasonable (dpl stuff) and how to it.

File: Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx:61]: The common wisdom was that the FEEIDs should be used to identify different links rather than  linkID's (which, apparently, may change). Not sure if this is relevant from the CTP.

File: Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx]: missing {} will be signaled by CI

File: Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx]: Instead of having dynamic object creation in a tight loop it would be better to define it outside of all loops and call `clear` here.

File: Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx]: {}

File: Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx]: I guess there should be `std::vector<gbtword80_t>& diglets` here.

File: Detectors/CTP/workflow/src/ReaderSpec.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/ReaderSpec.cxx]: From the filename is it not clear that this is a digit reader, better name it explicitly as DigitReaderSpec.
Also, note that we started to move gradually all readers and writers to separate IO directories, in order to avoid circular dependencies in global workflows.
Perhaps you could place this class directly in the CTP/WorkflowIO .

File: Detectors/CTP/workflow/src/ReaderSpec.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/ReaderSpec.cxx]: please remove commented code

File: Detectors/CTP/workflow/src/RecoWorkflow.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/RecoWorkflow.cxx]: why is this here?

File: Detectors/CTP/workflow/src/WriterSpec.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/WriterSpec.cxx]: Also this one better to rename to DigitWriterSpec and to have it in the WorkflowIO

File: Detectors/CTP/workflow/src/ctp-reco-workflow.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/ctp-reco-workflow.cxx:19]: In order to create a correct DPL `DataHeadr.firstTForbit` in the root file digits driven mode , the workflows should have the lines like https://github.com/AliceO2Group/AliceO2/blob/59df0714e1705fb6e3d55b8f0bda869f2a2b866f/Detectors/ITSMFT/ITS/workflow/src/its-reco-workflow.cxx#L15 
https://github.com/AliceO2Group/AliceO2/blob/59df0714e1705fb6e3d55b8f0bda869f2a2b866f/Detectors/ITSMFT/ITS/workflow/src/its-reco-workflow.cxx#L40
https://github.com/AliceO2Group/AliceO2/blob/59df0714e1705fb6e3d55b8f0bda869f2a2b866f/Detectors/ITSMFT/ITS/workflow/src/its-reco-workflow.cxx#L72-L73

So, add here `#include "DetectorsRaw/HBFUtilsInitializer.h"`

File: Detectors/CTP/workflow/src/ctp-reco-workflow.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/ctp-reco-workflow.cxx]: add here `o2::raw::HBFUtilsInitializer::addConfigOption(options);`

File: Detectors/CTP/workflow/src/ctp-reco-workflow.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/ctp-reco-workflow.cxx]: change this to 
````
auto wf = o2::ctp::reco_workflow::getWorkflow( ... );

// configure dpl timer to inject correct firstTFOrbit: start from the 1st orbit of TF containing 1st sampled orbit
o2::raw::HBFUtilsInitializer <span class="pl-smi">hbfIni</span>(configcontext, wf);
return std::move(wf);
````

**Review by shahor02 — CHANGES_REQUESTED** [review]
Hi @lietava 

The dpl data fetching part looks ok, see few comments below.

File: Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx
Comment (lietava) [reply Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx:61]: I see the point. But that would only work if our firmware would add FEEID to RDH, right ?

File: Detectors/CTP/workflow/src/RecoWorkflow.cxx
Comment (lietava) [reply Detectors/CTP/workflow/src/RecoWorkflow.cxx]: remnant of copy and paste. removed.

File: Detectors/CTP/workflow/src/WriterSpec.cxx
Comment (lietava) [reply Detectors/CTP/workflow/src/WriterSpec.cxx]: I rename. I do not understand second part of sentence: " to have it in the WorkflowIO" ?

File: Detectors/CTP/workflow/src/ctp-reco-workflow.cxx
Comment (lietava) [reply Detectors/CTP/workflow/src/ctp-reco-workflow.cxx:19]: I added include and also equivalents of lines 40 and 73 ?

Comment (lietava) [reply Detectors/CTP/workflow/src/ctp-reco-workflow.cxx:19]: ok, your next two comments resolving this.

**Comment by lietava** [comment]
@shahor02 1.) using fee instead of link - I'll do it later. 2.) I am not sure I fully understand your comment on (Digit)WriterSpec - please see above. All other comments implemented, I think/ Cheers, R.

File: Detectors/CTP/workflow/src/WriterSpec.cxx
Comment (shahor02) [reply Detectors/CTP/workflow/src/WriterSpec.cxx]: I mean put reader and writer workflows in a separate library, depending only on the data formats. E.g. this can be WorkflowIO directory creating a library of the same name. Or, see how TRD did this, creating `io` directory in the `O2/Detectors/TRD/workflow`

File: Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx
Comment (shahor02) [reply Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx:61]: Afaik, the FEEID is supposed to be unique and fixed in all cases.

**Comment by lietava** [comment]
Hi @shahor02 , I created workflowIO on the same level as workflow. I was not sure where to put CTPDigitWriterSpec - kept it in workflow.
Anything else apart links (later) and missing STF (can do in this PR or next) ?
Cheers, Roman.

File: Detectors/CTP/workflowIO/src/ReaderSpec.cxx
Comment (shahor02) [inline Detectors/CTP/workflowIO/src/ReaderSpec.cxx]: you still have this stray lines.

File: Detectors/CTP/workflowIO/src/ReaderSpec.cxx
Comment (shahor02) [inline Detectors/CTP/workflowIO/src/ReaderSpec.cxx]: I thought you will rename this to DigitReaderSpec, to be symmetric with Writer.

**Review by shahor02 — COMMENTED** [review]
Hi @lietava 

> I was not sure where to put CTPDigitWriterSpec - kept it in workflow.

Just noticed: why do you need both CTPDigitWriterSpec and DigitWriterSpec? They are doing exactly the same, except that in the CTPDigitWriterSpec you already have an option of indicating that it is called from the digitization or decoding. 
I would eliminate the DigitWriterSpec and move the CTPDigitWriterSpec to workflowIO.

Apart from that, the code seems to be complete, but did you try to decode simulated raw data and make sure that the decoded digits exactly match to simulated ones?

Once this is validated and the PR is merged I will implement the CTF creation and add the CTP to the FST.

**Comment by lietava** [comment]
Hi @shahor02 , I did all modifications you suggested above. 
Also I tried to implement no lost TF trick - please, check.
If ok - I can test rawtodigit -  so I run rawtodigit and then you have some tool to check  ?
Cheers, R.

File: Detectors/CTP/workflow/include/CTPWorkflow/RawToDigitConverterSpec.h
Comment (shahor02) [inline Detectors/CTP/workflow/include/CTPWorkflow/RawToDigitConverterSpec.h]: I think CI will require to change this to `RawToDigitConverterSpec() = default`

File: Detectors/CTP/workflow/include/CTPWorkflow/RawToDigitConverterSpec.h
Comment (shahor02) [inline Detectors/CTP/workflow/include/CTPWorkflow/RawToDigitConverterSpec.h]: This should be removed

File: Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx:51]: See this comment: https://github.com/AliceO2Group/AliceO2/pull/6457#discussion_r653807498
You should push an empty output in this case.

File: Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx]: not needed

**Review by shahor02 — COMMENTED** [review]
Hi @lietava 
Looks fine but see some comments below. For the validation: I don't have tools for that, one should simply create raw data from simulated digits, then run your new decoding, write digits to other file and make sure that the content of both digit files are the same.

File: Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx
Comment (lietava) [reply Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx:51]: added output, pls, check

**Comment by lietava** [comment]
HI @shahor02 , I did fixes you suggested, please,  check. For validation I involve Marek. Merging before or after validation - up to you.

**Comment by lietava** [comment]
@shahor02 . I tried quick check and got:
[O2/latest-ctpdev-o2] ~/o2tests $> o2-ctp-reco-workflow -b
[ERROR] invalid workflow in o2-ctp-reco-workflow: No matching output found for <matcher query: (and origin:CTP (and startTime:$0  description:RAWDATA ))> as requested by data processor "CTP-RawStreamDecoder". Candidates:
-CTP/DIGITS/0

File: Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx
Comment (shahor02) [inline Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx]: you already have `mOutputDigits.clear();` in the beginning.

**Review by shahor02 — COMMENTED** [review]
Hi @lietava ,

As I wrote before, you should use `o2-raw-file-reader-workflow` to inject the raw data into the DPL : https://github.com/AliceO2Group/AliceO2/tree/dev/Detectors/Raw#raw-data-file-reader-workflow.
This means that having generated raw data, you should do e.g.
````
o2-raw-file-reader-workflow --input-conf raw/CTP/CTPraw.cfg | o2-ctp-reco-workflow 
````

**Comment by lietava** [comment]
> o2-raw-file-reader-workflow --input-conf raw/CTP/CTPraw.cfg | o2-ctp-reco-workflow

**Comment by lietava** [comment]
> > o2-raw-file-reader-workflow --input-conf raw/CTP/CTPraw.cfg | o2-ctp-reco-workflow

@shahor02  I did enter enviroment ...

[O2/latest-ctpdev-o2] ~/o2tests $> o2-raw-file-reader-workflow --input-conf CTPraw.cfg | o2-ctp-reco-workflow 
Error 65542: GLX: GLX version 1.3 is required
o2-ctp-reco-workflow: /home/rl/alice/sw/SOURCES/GLFW/3.3.2/3.3.2/src/input.c:863: glfwSetMouseButtonCallback: Assertion `window != ((void *)0)' failed.
Aborted

**Comment by lietava** [comment]
> > > o2-raw-file-reader-workflow --input-conf raw/CTP/CTPraw.cfg | o2-ctp-reco-workflow
> 
> @shahor02 I did enter enviroment ...
> 
> [O2/latest-ctpdev-o2] ~/o2tests $> o2-raw-file-reader-workflow --input-conf CTPraw.cfg | o2-ctp-reco-workflow
> Error 65542: GLX: GLX version 1.3 is required
> o2-ctp-reco-workflow: /home/rl/alice/sw/SOURCES/GLFW/3.3.2/3.3.2/src/input.c:863: glfwSetMouseButtonCallback: Assertion `window != ((void *)0)' failed.
> Aborted

@shahor02 Sorry, it was just gui. Using -b  makes it run.

**Comment by lietava** [comment]
@shahor02 . Thanks. It produces some digits now. As I said I suggest Marek with me starts to play with it.
Shall we merge to dev before ?
R.

**Comment by shahor02** [comment]
If it does not break the stability of the code, we can merge it, though is there a problem to validate it before?
In any case, before merging you should remove the Draft tag from the PR and it should pass the CI.

**Comment by lietava** [comment]
@shahor02 ok, we validate first.

**Comment by lietava** [comment]
@shahor02 I was trying today to rebase dev to ctpdev  but it fails as there is renamed file I guess.
The validation may take weeks as I am not doing it now with full capacity and Marek needs some learning.
So my question is if it is not worth to merge at this stage just to reduce divergence. But if it is ok to stay like this I am fine with it. Decision up to you. Cheers, R.

**Comment by shahor02** [comment]
Hi @lietava 
I would not merge the code which is in draft mode. So, remove the draft tag so that CI could start.

For the rebasing: yes, it complains on the conflict between the CTPDigitWriterSpec modified in the dev and removed in your branch. You should do (after updating your dev):
````
git rebase dev
# will complain on Detectors/CTP/workflow/src/CTPDigitWriterSpec.cxx, Detectors/CTP/workflow/include/CTPWorkflow/CTPDigitWriterSpec.h

git rm Detectors/CTP/workflow/src/CTPDigitWriterSpec.cxx Detectors/CTP/workflow/include/CTPWorkflow/CTPDigitWriterSpec.h
git rebase --continue
````

**Comment by shahor02** [comment]
@lietava could you rebase and resolve the conflicts? The tests will not start w/o that.

**Comment by lietava** [comment]
Conflict resolved.

**Comment by lietava** [comment]
Looking in log build/AliceO2/O2/o2/macOS : are these erros because of CTP ?

**Comment by shahor02** [comment]
@lietava no, macOS is broken in all PRs.

**Comment by lietava** [comment]
I fixed license issues and now I do not see any CTP related issues in two failed logs. Please, let me know I am wrong.

**Comment by shahor02** [comment]
right, they are not related. What about validation?

**Comment by lietava** [comment]
@shahor02 waiting for Marek. I am busy with other things but may reorder. Is it urgent ? R.

**Comment by shahor02** [comment]
no, just to understand if I shall merge this or not. Since this PR does not break anything, I could merge it.

**Comment by lietava** [comment]
@shahor02 , please, merge. FYI: I did (de)coder checks outside O2 so bug maybe somewhere in creating HBF. Then I also like to involve Marek - other purpose of exercise. In parallel I start to add counters infrastructure. Cheers, R.

**Pull request merged**
